### ISSUE TITLE
Give coLunacyDNS IPv6 Windows socket support

### ISSUE THREAD

**Issue by samboy** [samboy/MaraDNS#80]
Back in 2007, I was able to give Deadwood for Linux hybrid working IPv4 + IPv6 sockets, while also having the socket code (with some #ifdefs) also compile with Windows IPv4 sockets via `mingw` in Windows. However, the Windows sockets were IPv4 only. This wasn’t that big of a deal back in 2007, when IPv6 was just starting to roll out, home ISPs only offered, IPv4, and when IPv4 addresses were still readily available. Here, in 2020, IPv4 addresses have already run out in a lot of regions of the world, my home ISP has given me an IPv6 /64, and about 1/3 of the traffic Google receives is IPv6 traffic.

In addition, Windows expects to see a DNS server with IPv6 in its IPv6 configuration.

That in mind, it’s time to make a DNS server with full Windows and IPv6 support. I will limit this effort to coLunacyDNS, just so Windows users on IPv6 can use it as a 6-to-4 gateway for DNS. DNS servers are still by and large using IPv4, and it’s going to be a while before recursive DNS servers start performing IPv6 lookups when given a glueless NS referral.

**Comment by samboy** [comment]
It looks like even my ancient version of mingw from 2003 has INET6 sockets in its version of `winsock.h`. There is an `AF_INET6` there. In addition `ws2tcpip.h` does have `sockaddr_in6`.

Some Microsoft docs: https://docs.microsoft.com/en-us/windows/win32/winsock/ipv6-enabled-server-code-2

But this may be a simple matter of just adding `ws2tcpip.h` as a header file and doing the hybrid sockets JJS and I set up.

It looks like we do *not* have inet_pton, so I will have to write my own parser, or I will have to force Windows users to use a different format for IPv6 IPs (the `2001-0db8-0f00-0ba4 0000-0000-0000-1234` form).

**Comment by samboy** [comment]
The ancient version of MinGW I develop against to make Windows32 binaries doesn’t have `inet_pton` for converting human readable strings in to IPv6 addresses, so I will have to implement this (again).  See https://github.com/samboy/ip6Parse

**Comment by samboy** [comment]
That `ip6Parse` library is nicely rounded out, with the ability to support a superset of standard IPv6 IPs, both standard colon-separated IPs and “MaraDNS” formatted IPs.  Next: Add it to coLunacyDNS.

**Comment by samboy** [comment]
Done. MaraDNS 3.5.0013 is released and has a Windows32 binary of coLunacyDNS with IPv6 support. I was able to compile a service which binds to an IPv6 socket using an old Windows XP virtual machine and a 2003 version of MinGW/Msys (I didn’t think IPv6 sockets were around back then, but they were and still work today). I was under the impression ancient mingw didn’t have IPv6 because it doesn’t have `inet_pton`; however, it has *everything else* to handle IPv6 sockets, so I spent a day writing my own IPv6 address parser.

There are some rough edges: We have to use a MaraDNS-specific format for IPv6 addresses when specifying IPv6 addresses to return to the client (we can use standard IPv6 addresses when specifying the bind address, as well as MaraDNS-specific ip6 addresses, so the code is there in coLunacyDNS). We have a lot of unneeded `NOIP6` #ifdefs in the code. The Windows coLunacyDNS text document should be updated (it only briefly touches on IPv6 binding). But, by golly, MaraDNS now has a Windows 32 binary which can use IPv6 sockets.
